New fsinfo channel

allisonkarlitskaya opened the pull request:
- [x] ~#19842~ no longer depends on this
 - [x] implement less-adhoc file content size limitation (good enough for now — we can add an option if someone requests it)
 - [x] return problem code in close message on error
 - [x] once-over look at the channel implementation and docs
 - [x] really quite a lot of unit tests 

Left for follow-ups:
 - what to do about the case where `path` is a directory which gains/loses read/search permission while we're watching it
 - add support for reporting symlink target attributes (possibly without monitoring for changes)
 - maybe some kind of `file(1)`-flavour content identification attribute, mime type sniffing, etc.  gio has this.

cockpituous commented on pkg/lib/cockpit.js:
```diff
@@ -3700,6 +3721,41 @@ function factory() {
+
+        function on_message(_event, payload) {
+            const msg = JSON.parse(payload);
+            self.value = json_merge(self.value, msg);
```
This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19845-20240115-145734-3653cecb-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on pkg/lib/cockpit.js:
```diff
@@ -3700,6 +3721,41 @@
+            const msg = JSON.parse(payload);
+            self.value = json_merge(self.value, msg);
+            for (const callback of callbacks) {
+                callback(self.value);
```
This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19845-20240115-145734-3653cecb-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on pkg/lib/cockpit.js:
```diff
@@ -3700,6 +3721,41 @@
+            }
+        }
+
+        self.effect = function(callback) {
+            callback(self.value);
```
These 2 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19845-20240115-145734-3653cecb-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on pkg/lib/cockpit.js:
```diff
@@ -3700,6 +3721,41 @@
+
+        const channel = cockpit.channel({
+            payload: 'fsinfo',
+            superuser: 'try',
+            'close-on-error': false,
+            watch: true,
+            path,
+            attrs,
+            ...options
```
These 6 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19845-20240115-145734-3653cecb-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on pkg/lib/cockpit.js:
```diff
@@ -3700,6 +3721,41 @@
+            attrs,
+            ...options
+        });
+        channel.addEventListener("message", on_message);
```
This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19845-20240115-145734-3653cecb-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on pkg/lib/cockpit.js:
```diff
@@ -3700,6 +3721,41 @@
+        });
+        channel.addEventListener("message", on_message);
+
+        return self;
```
This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19845-20240115-145734-3653cecb-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

allisonkarlitskaya commented:
Right.  The giant wall of rhel/centos failures is a good reminder that we can't really move `cockpit.js` to use this stuff yet, until we drop the C bridge...

cockpituous commented on pkg/lib/cockpit.js:
```diff
@@ -3700,6 +3719,41 @@ function factory() {
+        function on_message(_event, payload) {
+            const msg = JSON.parse(payload);
+            self.value = json_merge(self.value, msg);
+            for (const callback of callbacks) {
```
This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19845-20240116-103228-adaee4e9-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on pkg/lib/cockpit.js:
```diff
@@ -3700,6 +3719,41 @@
+            }
+        }
+
+        self.effect = function(callback) {
+            callback(self.value);
+            callbacks.push(callback);
```
These 3 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19845-20240116-103228-adaee4e9-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on pkg/lib/cockpit.js:
```diff
@@ -3700,6 +3719,41 @@
+        const channel = cockpit.channel({
+            payload: 'fsinfo',
+            superuser: 'try',
+            'close-on-error': false,
+            watch: true,
+            path,
+            attrs,
+            ...options
```
These 5 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19845-20240116-103228-adaee4e9-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on pkg/lib/cockpit.js:
```diff
@@ -3700,6 +3719,41 @@
+            attrs,
+            ...options
+        });
+        channel.addEventListener("message", on_message);
```
This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19845-20240116-103228-adaee4e9-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on pkg/lib/cockpit.js:
```diff
@@ -3700,6 +3719,41 @@
+        });
+        channel.addEventListener("message", on_message);
+
+        return self;
```
This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19845-20240116-103228-adaee4e9-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

allisonkarlitskaya commented on src/cockpit/channels/filesystem.py:
```diff
@@ -268,3 +287,241 @@
+
+        available_stat_getters = {
+            'type': lambda buf: buf and stat_types.get(stat.S_IFMT(buf.st_mode), 'special'),
+            'tag': lambda buf: tag_from_stat(buf),
```
fair enough.  Thanks, bot!

martinpitt commented on doc/protocol.md:
```diff
@@ -829,6 +829,119 @@ or pipe is done.
+ * `path` (mandatory): a string
+ * `attrs` (mandatory): a list of strings, the list of attributes to be
+   reported for the requested file.  See below for the possibilities.
+ * `fnmatch`: (optional, default ''): if non-empty, this is a [`fnmatch`-style
```
Just to avoid doubt:
 - that default means that if `path` is a directory, this won't report any information about the contents, just the dir itself.
 - if `path` is not a directory, this option gets ignored?

This could be clarified.

allisonkarlitskaya replied:
very precisely: the requested information about the file (whatever it is) is always reported.  Additionally, if `fnmatch` is non-empty, and `path` happens to point to a directory at any given time, and we can open that directory for reading, then there will be an `entries` attribute added: a dictionary mapping filenames to objects describing those files.

If `path` is not a directory, or if `path` can't be opened for reading, there will be no `entries`.  If the directory can be opened for reading but is empty (excluding `.` and `..`) then `entries` will be present with a value of `{}`.

This is very similar to requesting the "content" and "target" attributes.  They are only reported if the file type is correct, otherwise they are missing.

Originally you opted into having "entries" reported by having "entries" in the attr array, which was a more logically consistent API.  I found this to be a bit awkwards in use and implementation though, and it led to a lot of complex argument checking.  The new thought it: if we have to have `fnmatch` anyway, this is a far better way to get the same behaviour.

martinpitt commented on doc/protocol.md:
```diff
@@ -829,6 +829,119 @@ or pipe is done.
+   (with `problem`) on failure to `open()` the named `path`
+ * `watch` (optional, default: false): if the channel should watch for changes
+ * `follow` (default: true): if the trailing component of `path` is a symbolic
+   link, whether we should follow it.  `no-follow` mode is not supported for
```
`no-follow` isn't an option name or value AFAICS, so either make this `follow: false` or "no-follow" (without quotes)

allisonkarlitskaya replied:
Thanks.  This got renamed.  I didn't like the double negative of the default value being `"no-follow": false`.

martinpitt commented on doc/protocol.md:
```diff
@@ -829,6 +829,119 @@ or pipe is done.
+   link, whether we should follow it.  `no-follow` mode is not supported for
+   `watch` (but is also mostly unnecessary).
+
+A channel is created with an associated absolute pathname ('path').  A path is
```
`path` ? (with backticks), to refer to the option

allisonkarlitskaya replied:
Sure.

martinpitt commented on doc/protocol.md:
```diff
@@ -829,6 +829,119 @@ or pipe is done.
+   pattern](https://docs.python.org/3/library/fnmatch.html) selecting the directory
+   entries to be reported.  Use `*` to report all files.  Can also be used to
+   implement search (`searc*`) or hide temporary files (`[!.]*`).
+ * `close-on-error` (optional, default: true): if the channel should close
+   (with `problem`) on failure to `open()` the named `path`
```
We already discussed this a week ago and you told me, but I keep tripping over this: what happens with `false`? It gets explained two pages down (and not super easy to find). Perhaps this should just say "see below".

allisonkarlitskaya replied:
I'll add a "see below".  Keep in mind that this is `protocol.md`, which is basically a reference guide for ourselves...

martinpitt replied:
This is gone -- this was confusing and also mostly redundant with `watch`.

martinpitt commented on doc/protocol.md:
```diff
@@ -829,6 +829,119 @@ or pipe is done.
+
+A channel is created with an associated absolute pathname ('path').  A path is
+a string that you pass to the `open()` syscall.  The object returned by that
+call is described by the (JSON Document) value of the channel.
```
Why the parens?

allisonkarlitskaya replied:
The sentence parses with or without the content of the parens, and the content adds extra optional information, which is what parens are usually for.  I can remove them if you prefer.

martinpitt commented on doc/protocol.md:
```diff
@@ -829,6 +829,119 @@ or pipe is done.
+
+Conceptually, the channel represents the value of a JSON object describing what
+is found at the path, or the reason for an error.  Each message on the channel
+is a JSON Merge Patch (RFC7396) which updates or replaces this object.
```
Link to https://datatracker.ietf.org/doc/html/rfc7396 ?

allisonkarlitskaya replied:
OK.

martinpitt commented on doc/protocol.md:
```diff
@@ -829,6 +829,119 @@ or pipe is done.
+The top-level keys in the object are:
+
+ - `info`: the info about the reported path (described below)
+ - `problem`: a "problem code" like `not-found` or `not-directory`
```
OOI, when would `not-directory` happen? Intuitively, with a non-empty `fnmatch` and `path` not being a dir?

allisonkarlitskaya replied:
There used to be a separate `onlydir` boolean argument, but I got rid of it in favour of having the path end with `/`, but forgot to document that.  Thanks for the catch.

martinpitt commented on doc/protocol.md:
```diff
@@ -829,6 +829,119 @@ or pipe is done.
+
+ - `info`: the info about the reported path (described below)
+ - `problem`: a "problem code" like `not-found` or `not-directory`
+ - `message`: a user-readable (`strerror`) string describing the error
```
:thinking: that might hint devs to show the string to the UI as-is, but (1) that's not translatable, and (2) putting error string comparisons into code is brittle and ugly. Before we commit to that API (which will be difficult to remove), perhaps just give the numeric `errno` or leave it out completely, and we wait until we have a good use case? Do you need this for navigator, or is it sufficient there to just look at `problem` or even just check for success/failure?

allisonkarlitskaya replied:
I thought about this, and almost wrote (English) into the text, but it's not true.  We'll get this in whatever the default language of the user running the bridge is.

In any case, we also report "problem" (and can expand the list of values we have there to match more error codes).

I'm against reporting errno, unless we report it in string form.  These numbers are architecture-specific and quite useless to JSworld.

martinpitt replied:
Ah, if they are already translated, then I have no objection. Perhaps mention that in the docs, to discourage devs from trying to match on it?

allisonkarlitskaya replied:
They're actually not already translated, but they theoretically should be.  This is another (rather old) topic about when/if/how we should call `setlocale()`.

allisonkarlitskaya replied:
I checked and it would be possible to report a symbolic errno code like `ENOENT` etc...

martinpitt replied:
With the availability of `errno` callers can ignore `message`, or only use it for `console.log` or so, that's good enough for me.

martinpitt commented on doc/protocol.md:
```diff
@@ -829,6 +829,119 @@ or pipe is done.
+
+If `close-on-error` is set, `problem` and `message` will never be sent, but the
+channel will be closed with the `problem` code set to indicate the error.
+
```
Please drop the empty line here; the "otherwise" refers to close-on-error.

allisonkarlitskaya replied:
OK.

martinpitt commented on doc/protocol.md:
```diff
@@ -829,6 +829,119 @@ or pipe is done.
+attributes of the file.  Any attribute which cannot be read (eg: due to
+permissions problems) is simply not reported.
+
+ * `type`: a string: `file`, `directory`, `link`, or `special`
```
To make sure it's clear: if `path` is a link, then the value depends on the `follow` option? and for directory contents it doesn't?

allisonkarlitskaya replied:
From further down in the doc:

> If the provided `path` is a symbolic link, it is followed, unless
> `follow: false` is specified.  If symbolic links are present in the directory,
> they are always reported directly (ie: not followed).

martinpitt commented on doc/protocol.md:
```diff
@@ -829,6 +829,119 @@ or pipe is done.
+permissions problems) is simply not reported.
+
+ * `type`: a string: `file`, `directory`, `link`, or `special`
+ * `mode`: an integer representing the "file permission" bits.  As per JSON,
```
Want to add `st_mode` here, like for the other fields below?

allisonkarlitskaya replied:
It's not exactly `st_mode`.  It should be `st_mode & 0o7777` aka `stat.S_IMODE(buf.st_mode)` (but is currently implemented as `0777`, which needs fixing).  I can add that, and I'll fix the implementation.

allisonkarlitskaya replied:
Done and done.

martinpitt commented on doc/protocol.md:
```diff
@@ -829,6 +829,119 @@ or pipe is done.
+   this is transmitted as a decimal value, but it is meant to be interpreted in
+   octal, in the usual way.
+ * `uid`: an integer, the uid of the file owner (`st_uid`)
+ * `owner`: a string, or an integer if the lookup failed
```
That's strange; are user names forbidden in principle to have only numeric characters? it's also inconsistent with "Any attribute which cannot be read is simply not reported". I'd rather leave it out, which makes the API more robust.

allisonkarlitskaya replied:
If it reports the username then it's a string.  If it reports the uid, it's an integer, which is done in order to make the two cases easily distinguishable.

You can get the raw uid by requesting `uid`.

The idea here is that a hypothetical UI will probably want to display the username if it's available, and fall back to displaying the numeric UID otherwise.  If we don't implement that fallback here, then the user needs to subscribe to both `uid` and `user` in order to do that, which is 99% of the time redundant.

We could discuss always returning a string here, and if that string is numeric then the caller knows that they got a UID (and that's probably what they wanted to directly display anyway).  I personally prefer it as an int.

martinpitt replied:
I'm just concerned as this is JavaScript, which doesn't tend to make too fine a difference between them. `Number.isInteger()` does the right thing. OTOH `<span>{ props.owner ?? props.uid }</span>` feels more explicit to me, and makes it easier for UIs to actually tell whether a UID is valid or not (they may be formatted differently, for example).

Subscribing to both uid and user seems okay to me, if a program actually needs/wants that (most won't -- navigator is of course a case where it matters). But there is no additional overhead to detecting the uid.

Strict no for "always return a string". If you really like the int/string dual type here, then I don't veto it, but I prefer my types clean. :grin:

allisonkarlitskaya replied:
Let's split the baby and introduce a third property!

 - `uid`: int
 - `user`: string or null
 - `uhh`: string or int

allisonkarlitskaya replied:
Corresponding attribute for `gid` being called `guhh`, obviously.

allisonkarlitskaya replied:
I guess my final word on this is: I don't know any situation where it would be useful to request `user` and *not* get the uid as a fallback in case the lookup fails, so I'm going to keep it the way it is.

martinpitt replied:
OK -- IMHO it makes the API harder and more brittle to use on the caller side, but at least it's possible.

martinpitt replied:
... in the sense of "requesting every single of those in `attrs` feels like overkill" -- you'd either request and get all ownership and permissions, or none of them, as they all come from a single stat() call anyway. I totally understand only getting dir contents and file contents upon request, but everything that comes from the same source should be treated as a property group.

martinpitt replied:
OK -- I can't say that a multi-typed value sparks joy in me from the API user POV, but it's good enough and it's your PR. I don't want to risk getting into baby splitting! :grin:

martinpitt commented on doc/protocol.md:
```diff
@@ -829,6 +829,119 @@ or pipe is done.
+ * `uid`: an integer, the uid of the file owner (`st_uid`)
+ * `owner`: a string, or an integer if the lookup failed
+ * `gid`: an integer, the gid of the file group (`st_gid`)
+ * `group`: a string, or an integer if the lookup failed
```
dito

martinpitt commented on doc/protocol.md:
```diff
@@ -829,6 +829,119 @@ or pipe is done.
+ * `gid`: an integer, the gid of the file group (`st_gid`)
+ * `group`: a string, or an integer if the lookup failed
+ * `size`: an integer, the (apparent) size of the file (`st_size`)
+ * `modified`: a float, the mtime of the file (`st_mtim`)
```
I buy an "e"

allisonkarlitskaya replied:
Nope.  `st_mtim` is the correct spelling here, if you want the sub-second information (which we do report).  `st_mtime` is an int.

martinpitt replied:
Ugh.. TIL!

martinpitt commented on doc/protocol.md:
```diff
@@ -829,6 +829,119 @@ or pipe is done.
+ * `modified`: a float, the mtime of the file (`st_mtim`)
+ * `tag`: a value type: the current 'transaction tag' of the file, with the
+   same meaning as elsewhere in this document.  Ideally: this changes if the
+   content of the file changes.  This is stable between different channels.
```
What does "stable between different channels" mean? fswatch, fsinfo, fsread will report the same tag for the same file state?

allisonkarlitskaya replied:
Yes.  Do you have a suggestion for better wording?

martinpitt replied:
hmm.. "This is the same as in fsread1 and fsreplace1"? But no strong opinion, feel free to resolve this. I mostly just wanted to check that I understood it right.

martinpitt commented on doc/protocol.md:
```diff
@@ -829,6 +829,119 @@ or pipe is done.
+   same meaning as elsewhere in this document.  Ideally: this changes if the
+   content of the file changes.  This is stable between different channels.
+ * `entries`: for directories: the contents of the directory (see below)
+ * `content`: for regular text files, the content (if less than 4096 bytes)
```
That feels a bit arbitrary and too rigid. Shouldn't the caller decide via `attrs` whether it's sensible to directly read the files (if they need it in a race-free manner) or not (even if they are small)? Cf.  https://cockpit-project.org/guide/latest/cockpit-file.html#cockpit-file-notify . For safety I'd rather see a `max_read_size` option like fsread1 has.

allisonkarlitskaya replied:
The caller does decide via `attrs` to read file content: by requesting the `content` attr.  It will never be reported if not requested.

I considered adding `max-read-size` but figured we could wait-and-see if it's actually needed.  I actually have another plan here, anyway: I do want to support large files, but I want to do it via opening an external channel somehow (which needs some thinking — we currently have no concept of "push channels").

That's why I decided to leave this one hanging for the time being...

martinpitt replied:
Hmm, so perhaps this PR should drop "content" altogether for now?

martinpitt replied:
It's gone now. This is a bit tricky, so let's introduce it in a follow-up if/when we get a consumer.

martinpitt commented on doc/protocol.md:
```diff
@@ -829,6 +829,119 @@ or pipe is done.
+ * `content`: for regular text files, the content (if less than 4096 bytes)
+ * `target`: for symbolic links: the target of the link (a string)
+
+The `entries` attribute is only reported on directories, and only if the
+contents of the directory can be read.  It is an object where each key is the
```
"... and only if `fnmatch` is set to a non-empty value" or similar? cf. my question above.

allisonkarlitskaya replied:
Yes.  This is a holdover from before `fnmatch` existed and `entries` was explicitly set via `attrs`, as the others.  I'll fix it.

martinpitt commented on doc/protocol.md:
```diff
@@ -829,6 +829,119 @@ or pipe is done.
+file, using the same attributes as above.  `entries` is not reported on entries
+(ie: no recursive listing).
+
+The attributes to be reported can be controlled by setting the `attrs` option.
```
s/can be/are/, as it's a mandatory option.

allisonkarlitskaya replied:
Sure.

martinpitt commented on doc/protocol.md:
```diff
@@ -829,6 +829,119 @@ or pipe is done.
+if `/usr`, `/etc`, or any other directories or intermediate links were changed,
+deleted, or moved out of the way.
+
+(TODO: providing `watch: true, follow: false` would be possible with an
+addition to the `systemd_ctypes` `PathWatch` API).
```
This doesn't belong into the docs; perhaps make it a comment in the code?

martinpitt commented on doc/protocol.md:
```diff
@@ -829,6 +829,119 @@ or pipe is done.
+   `close-on-error` is true, opening fails (`close` with `problem`).
+
+ - Otherwise, `ready` is sent and the value representing the current state is
+   reported, as above.  Only when the current value reaches a state where
+   `partial` is unset can the initial value be read.
```
I find this hard to understand. I would guess if partial is set, you need to continue reading messages until it is unset, then merge them all to get the final value?

allisonkarlitskaya replied:
Yes.  It's a bit complicated to explain this concept in general, but the idea is that the client is supposed to collect (and continuously merge) patches as they come off the wire, but only report the change to the API consumer once `partial` is not set.

In some sense it would make more sense from the viewpoint of the client to have `complete: true` to indicate that we have non-partial data, but in the current normal case we never set `partial` anyway, so we can avoid it entirely, which seemed nice.

allisonkarlitskaya replied:
I've explained this a bit more.

martinpitt replied:
Thanks, I find this much clearer now.

martinpitt commented on doc/protocol.md:
```diff
@@ -829,6 +829,119 @@ or pipe is done.
+   detected, they are sent as updates to the value of the channel.  If the path
+   itself cannot be opened at any point, and `close-on-error` is set, the channel
+   will close (with `problem`).  Otherwise, the value may become `null`, but
+   the channel will never close itself.
```
Will there ever be `partial` updates in watch mode? I guess/hope not, but this could be clarified here.

allisonkarlitskaya replied:
I'm not sure, and I intend to leave the question unanswered for now.

This is all going to be more meaningful if/when we try to split updates into smaller chunks to enable flow control.

allisonkarlitskaya replied:
There is one spot where `partial` will currently get used for watching, in fact:

If you transition from an error state to a directory then you'll send a patch like:

```js
{'problem': null, 'message': null, 'entries': { ... }}
```

and if you transition from directory to error:

```
{'problem': 'bzzt', 'message': 'uh oh', 'entries': null }}
```

but if you transition from one directory to another directory in the same location (as may happen if a symlink gets changed, for example) then you need to transition directly from reporting the attributes of one directory to reporting the attributes of a second directory.

That presents a problem.  Due to the way merge patching works, unless you want to explicitly send `"entries": {"foo": null, "bar": null}` updates for a potentially very large number of files, you can't replace the "entries" dict with a new one in one step.

For this case we send a message that sets `{"entries": null, "partial": true}` first, and then a second message which sets the desired end value of `"entries"`.

That's the existing implementation and it's a place where the client library should fail to report the intermediate data to its user.

martinpitt reviewed (changes requested):
Thanks! I'm really looking forward to this, it's been requested so many times. Also, I like how compact the implementation is.

I did an API and doc review. Most of it is trivial, some of it may need some discussion. Let's perhaps sort out the simple bits first, then I'm happy to have a high-bandwidth meeting to discuss contentious bits if necessary.

allisonkarlitskaya commented:
Changes:
 - lots of documentation fixups, clarifications, simplifications
 - dropped `close-on-error` mode, simplified accordingly
 - dropped `contents` attribute
 - changed from `file`, `directory`, `link`, `special` to the full spectrum of all 7 supported file types: `reg`, `dir`, `lnk`, `chr`, `blk`, `fifo`, `sock`.  The names are the "standard" ones from the `DT_*` constants for the `d_type` field of `struct dirent` as well as the `S_IF*` macros for `struct stat`'s `st_mode`.

allisonkarlitskaya commented:
Two more changes:
 * instead of reporting broken-out `problem` and `message` fields, we combine them into a `error` object and also add `errno` which is the symbolic name of an errno constant (like `ENOENT`, etc) which will give API consumers more to match on if they don't want to display the `message` directly
 * a minor adjustment to the semantics for sending resets to avoid a partial message in cases where it's possible

martinpitt commented on src/cockpit/channels/filesystem.py:
```diff
@@ -268,3 +287,226 @@ def do_close(self):
+        self.schedule_update(name)
+
+        if name and mask | (InotifyEvent.CREATE | InotifyEvent.DELETE |
+                            InotifyEvent.MOVED_TO | InotifyEvent.MOVED_FROM):
+            # These events change the mtime of the directory
```
Perhaps `and 'mtime' in attrs` (or a cached flag) as an optimization? As you were concerned about minimizing the events sent over the wire.

allisonkarlitskaya replied:
This would have no impact on what's sent over the wire: if mtime isn't in the attrs, we'll schedule the update, do the `fstat()` on the directory, notice that nothing has changed, and go back to sleep.

We could save ourselves the wake-up and the stat, I suppose... but we also just changed a file in the directory, so we're going to be waking up to stat the file, anyway.

The only place this would make a difference is if we were monitoring a directory, without monitoring its mtime, and without fnmatch set, and entries were changing a lot in the directory.  That seems like a fairly slim edge case.

Oh: and it's not just `mtime`.  The `tag` attribute also changes if the timestamp is modified.  Which is enough to convince me that adding a special check here would be far too brittle.  I can imagine we add some other mtime-dependent attribute in the future and forget to come back here to update it.

martinpitt commented on src/cockpit/channels/filesystem.py:
```diff
@@ -268,3 +287,226 @@ def do_close(self):
+
+    def do_open(self, options: JsonObject) -> None:
+        self.path = get_str(options, 'path')
+        if not os.path.isabs(self.path):
+            raise JsonError(options, '"path" must be an absolute path')
```
That's sensible as it avoids the assumption that the bridge runs in ~. However, should this support `os.path.expanduser()`, so that callers can watch e.g. `~/.ssh`? That seems useful to me. Totally fine as follow-up of course if/when we need it, just gauging your opinion.

allisonkarlitskaya replied:
I don't think we support this on any of the other filesystem channels...  My initial instinct here would to be complain about "what if the filename happens to start with `~`?", but indeed, it needs to start with `/`, so we're not in danger here...

If we wanted to add that as a feature, we should add it everywhere, and in a separate commit, I'd say.

martinpitt commented on src/cockpit/channels/filesystem.py:
```diff
@@ -268,3 +287,226 @@ def do_close(self):
+
+        else:
+            # PathWatch will essentially do the same as above
+            self.path_watch = PathWatch(self.path, self)
```
Note to self: on errors, PathWatch calls `do_identity_changed(None, errno)`. 5 points for Lissindor if the comment says that :grin:

allisonkarlitskaya replied:
Are you asking for more comments? For the five house points, I'd totally do it.

martinpitt replied:
Yes, pretty much what I  wrote, i.e. "will essentially do the same as above" → "calls do_identity_changed(None, errno) with the same behavior as above" or so

allisonkarlitskaya replied:
I'd like my 5 points, please.

martinpitt replied:
⭐ ⭐ ⭐ ⭐ ⭐

allisonkarlitskaya replied:
:heart:

martinpitt commented on test/pytest/test_bridge.py:
```diff
@@ -630,3 +635,282 @@ async def test_large_upload(event_loop, transport, tmp_path):
+
+
+def fsinfo_err(err: int) -> JsonObject:
+    problems = {
```
A bit sad to replicate report_error() here, but ok -- can't hurt.

allisonkarlitskaya replied:
I got sick of writing this out in full everywhere, particularly after the addition of the `errno` field...

martinpitt replied:
This wasn't meant to be an actionable thread, but if you want to factor it out I won't stop you :grin:  (But feel free to just close this thread).

martinpitt commented on test/pytest/test_bridge.py:
```diff
@@ -630,3 +635,282 @@ async def test_large_upload(event_loop, transport, tmp_path):
+    loopy.symlink_to(loopy)
+
+    expected_state: 'dict[Path, JsonObject]' = {
+        normal_dir: {"info": {"type": "dir", "entries": dir_entries}},
```
Would you mind expanding `dir_entries` here? It's easier to read, and avoids the definition so far away from this. It looks rather odd up there, as it's not part of the test setup, but the expected result.

martinpitt commented on test/pytest/test_bridge.py:
```diff
@@ -630,3 +635,282 @@ async def test_large_upload(event_loop, transport, tmp_path):
+
+            # generate some events immediately before we switch the link to
+            # ensure the event is suppressed
+            with contextlib.suppress(OSError):
+                from_file.touch()
+            with contextlib.suppress(OSError):
+                (from_file / 'a.txt').touch()
+            with contextlib.suppress(OSError):
+                (from_file / 'a.txt').unlink()
```
Why the suppresses here? That feels odd in a test -- first of all, they shouldn't fail (it's our tmpdir), and second it would just fail later on anyway as the test setup isn't what the assertions below expect?

allisonkarlitskaya replied:
`from_file` might not even be a directory... or we might not have write access to `touch()` it — we remove permissions from some of the files.

martinpitt replied:
Hmm, I'd prefer some explicit `.startswith('no-r')` or so to avoid surprises, but again: no strong opinion, feel free to just resolve this thread.

allisonkarlitskaya replied:
i made the comment a bit clearer about why we expect the failures.

martinpitt reviewed (changes requested):
I'm happy with the API now, and resolved all of my previous review threads. Thanks!

I reviewed the code now. :clap: for a real Lis-quality great code architecture! I have a few small niggles and questions, hopefully quick to sort out. Thanks!

allisonkarlitskaya commented on src/cockpit/channels/filesystem.py:
```diff
@@ -268,3 +287,226 @@ def do_close(self):
+        stat_types = {stat.S_IFREG: 'reg', stat.S_IFDIR: 'dir', stat.S_IFLNK: 'lnk', stat.S_IFCHR: 'chr',
+                      stat.S_IFBLK: 'blk', stat.S_IFIFO: 'fifo', stat.S_IFSOCK: 'sock'}
+        available_stat_getters = {
+            'type': lambda buf: buf and stat_types.get(stat.S_IFMT(buf.st_mode)),
```
This is a leftover from a time when we still ran the list if the stat failed (with buf as None).  I'll drop it.

allisonkarlitskaya commented on src/cockpit/channels/filesystem.py:
```diff
@@ -268,3 +287,226 @@ def do_close(self):
+            except FileNotFoundError:
+                return None
+            except OSError:
+                return {name: None for name, func in stat_getters}
```
The top-level "error" is only if we fail to open the item referred to by the `path` option on the channel.  Failing to stat a single item in the directory is reported by returning `{}` for that item.

A side note about `{}` vs a dictionary full of `None`: you correctly identify that this is a patch which deletes all of the keys.  To be more explicit: the way the code is structured, the `getattr()` function is expected to return the full list of attributes every time.  That means that if there are attributes which we can't report, we need to return them as `None` to make sure they get removed from the existing value.

Should indeed be `_func` or `_`.

martinpitt replied:
I thought `self.getattrs()` would be used for `path` itself as well. If these errors are handled somewhere else, then returning `{}` (effectively) makes sense, but then I'd say we should log these as warnings -- once we opened the directory fd, it should be really exceptional that fstat() or lstat fails, as all of these entries are in the directory, not related to the file inode, no?

allisonkarlitskaya replied:
`self.getattrs()` is used for the file itself, but only after the `O_PATH` fd has been successfully opened.  We call `fstat()`, which has no realistic chance to fail. Failures here are really about the subitems.

Note: we call this function on *any* change for the subitems, including deletion.  That's why we handle `FileNotFoundError` in a special way compared to all the other errors.

Other failures on statting items in the directory are indeed likely to be uncommon, but there's one possible case: a directory which has read permission but not search.  In that case, we'll see the directory entries, but won't be able to say anything about them.  That will result in `{}` being returned for each item, but at least the items will be listed.

It occurs to me that I didn't think about what should happen if search permission gets removed on a directory as we're watching it.  That's tricky.  Should we replace all of the existing files with `{}`?

And what if read permission gets removed?  Will inotify automatically shut down our watch in that case?

This probably could use a bit of investigation...

For the record, `ls` output in this case is the one where you see:

```
⬢[lis@toolbox cockpit]$ chmod a-x main; ls -l main
ls: cannot access 'main/containers': Permission denied
ls: cannot access 'main/doc': Permission denied
ls: cannot access 'main/examples': Permission denied
...
total 0
-????????? ? ? ? ?            ? aclocal.m4
-????????? ? ? ? ?            ? AUTHORS
-????????? ? ? ? ?            ? autogen.sh
d????????? ? ? ? ?            ? autom4te.cache
l????????? ? ? ? ?            ? bots
...
```

ie: it lists the files, but you can't see anything about them, relatively consistent with our result.

One difference between us and `ls` is that ls successfully identifies the file type still — that's surely a result of it using the `d_type` field of the dirent struct.  We could do that if we used `os.scandir()` instead of `os.listdir()`, but it would only work on the first iteration, which is why we favour just using `stat()`.

allisonkarlitskaya replied:
I think I have a plan for what we could do here — effective treat the change in read/search permissions on the directory that we're monitoring as a identity-change-level type of event and rescan everything from new.

Detecting the change in the case that we start out with the read permission is not difficult.  Detecting the other direction is a bit of a different story, though, as inotify will prevent us from adding a watch to begin with if we can't read.

In any case: I think this needs to (and certainly can) be punted to a follow-up

allisonkarlitskaya commented on src/cockpit/channels/filesystem.py:
```diff
@@ -268,3 +287,226 @@ def do_close(self):
+
+            result = {key: func(buf) for key, func in stat_getters}
+
+            if 'target' in result and stat.S_IFMT(buf.st_mode) == stat.S_IFLNK:
```
If we don't support a particular attribute, we still report it as None (due to the way we construct the `stat_getters` — if we don't know the attribute, we return None.  So the keys of `result` are exactly the requested attributes.  Checking a dictionary instead of checking a list (which is maybe more logical) of a pure performance thing.

Note: having the None values here helps us because it means we don't need to (for example) explicitly set 'target' to None in case of the file not being a symlink or the syscall failing.  That just keeps things a bit tidier below.

martinpitt replied:
Ah, now I get it -- you need to explicitly ask for `target` in `attrs`.

allisonkarlitskaya commented on src/cockpit/channels/filesystem.py:
```diff
@@ -268,3 +287,226 @@ def do_close(self):
+        assert self.fd is not None
+
+        entries: JsonDict = {name: self.getattrs(self.fd, name) for name in updates}
+        info = entries.pop('', {})
```
Sure.  The logic for this is that a lot of syscalls accept empty path to be "the dir itself" via the `AT_EMPTY_PATH` flag.  `readlinkat()` even accepts `''` without that flag, and we reply on that below.  The empty string is also what gets reported from inotify for changes to the directory itself.

Unfortunately, there's no way to use `AT_EMPTY_PATH` from Python that I could find, so we still have to do lstat() vs. fstat() for that part.  Maybe some day...

allisonkarlitskaya commented on src/cockpit/channels/filesystem.py:
```diff
@@ -268,3 +287,226 @@ def do_close(self):
+
+        entries: JsonDict = {name: self.getattrs(self.fd, name) for name in updates}
+        info = entries.pop('', {})
+        assert isinstance(info, dict)  # fstat() will never fail with FileNotFoundError
```
We've just extracted `info` as a value from the `entries` dictionary we built above.  The values in that dictionary come from the result of the `self.getattrs()` call, which returns a dictionary or None.  None is only returned in the case of `FileNotFoundError`, which will never be true for the root path itself (since we call `fstat()` in that case, not `lstat()` and `fstat()` can't fail with `ENOENT`).

The logic there is all fine and good, but mypy is (understandably) unable to infer that via the types alone, so we use this assert to help with type narrowing.  Otherwise, when we treat `info` as a dict below (by assigning to it) mypy would warn that it might be `None`.

I'm not thrilled that this was necessary and I tried a few other ways to get around this, but ultimately the requirement to have this here is fairly understandable...

martinpitt replied:
ah, so this is a mypy nudge, thanks.

allisonkarlitskaya commented on src/cockpit/channels/filesystem.py:
```diff
@@ -268,3 +287,226 @@ def do_close(self):
+            return True
+        else:
+            # only report updates on entry filenames if we match them
+            return fnmatch.fnmatch(name, self.effective_fnmatch)
```
We'll arrive here very often for directories, even if we didn't set an fnmatch.  systemd_types gives no control over the events that we monitor on the `path` that we watch, so even if we're only interested in monitoring a directory (and not its child items) we're going to get a lot of create/delete/etc. events.  This function is the very thing that filters them — and an empty fnmatch indeed matches nothing.

I had a special case here before to specifically check for empty string and return fast, but I decided the code looked nicer this way.  I didn't measure performance either way, but I don't imagine it would be a significant difference.

martinpitt replied:
OK, then the `and self.effective_fnmatch` would make this conceptually cleaner, and possibly also faster. I don't insist though, so feel free to just resolve if you want to keep it as-is. Thanks for checking!

allisonkarlitskaya commented on src/cockpit/channels/filesystem.py:
```diff
@@ -268,3 +287,226 @@ def do_close(self):
+            return False
+
+        buf = os.fstat(fd)  # this should never fail
+        if stat.S_IFMT(buf.st_mode) != stat.S_IFDIR:
+            self.report_error(errno.ENOTDIR)
+            return True
```
From a purely logical standpoint: what you say is true, but only with `follow: false`.  Otherwise, we'll have resolved the path to a directory already in `open()` and we'll have statted the directory here.  If `follow: false` *is* specified and the `path` is a symlink then of course `/` into that symlink will fail, and that's the desired result.

From a more pragmatic standpoint: this code path only hits on `watch` mode (because of the limitation in systemd_ctypes mentioned in the comment above).  `watch` mode is currently incompatible with `follow: false`, so we're never going to have an fd pointing to a symlink here.

allisonkarlitskaya commented on src/cockpit/channels/filesystem.py:
```diff
@@ -268,3 +287,226 @@ def do_close(self):
+        self.process_update({e for e in entries if self.interesting(e)}, reset=True)
+
+    def do_inotify_event(self, mask: InotifyEvent, cookie: int, rawname: 'bytes | None') -> None:
+        logger.debug('do_notify_event(%r, %r, %r)', mask, cookie, rawname)
```
Thanks :)

martinpitt reviewed (approved):
Thanks! Assuming green, I'm happy for this to land now. It'll likely get some follow-ups as the navigator side progresses.
